pkg/controller/instanceset/instance_util.go

```diff
@@ -236,18 +241,67 @@ func GenerateAllInstanceNames(parentName string, replicas int32, templates []Ins
 }
 
-func GenerateInstanceNamesFromTemplate(parentName, templateName string, replicas int32, offlineInstances []string) []string {
-	instanceNames, _ := GenerateInstanceNames(parentName, templateName, replicas, 0, offlineInstances)
+func GenerateAllInstanceNamesWithTemplatesIndexRanges(parentName string, replicas int32, templates []InstanceTemplate, offlineInstances []string, templatesIndexRanges []workloads.InstanceTemplateIndexRanges) ([]string, error) {
+	totalReplicas := int32(0)
+	instanceNameList := make([]string, 0)
+	for _, template := range templates {
+		replicas := template.GetReplicas()
+		indexPoints := make([]int32, 0)
+		var err error
+		if len(templatesIndexRanges) > 0 {
+			indexPoints, err = GetIndexPointsByTemplateName(templatesIndexRanges, template.GetName())
+			if err != nil {
+				return nil, err
+			}
+		}
+		names := GenerateInstanceNamesFromTemplate(parentName, template.GetName(), replicas, offlineInstances, indexPoints)
+		instanceNameList = append(instanceNameList, names...)
+		totalReplicas += replicas
+	}
+	if totalReplicas < replicas {
+		indexPoints := make([]int32, 0)
+		var err error
+		if len(templatesIndexRanges) > 0 {
+			indexPoints, err = GetIndexPointsByTemplateName(templatesIndexRanges, "")
+			if err != nil {
+				return nil, err
+			}
+		}
+		names := GenerateInstanceNamesFromTemplate(parentName, "", replicas-totalReplicas, offlineInstances, indexPoints)
+		instanceNameList = append(instanceNameList, names...)
+	}
+	getNameNOrdinalFunc := func(i int) (string, int) {
+		return ParseParentNameAndOrdinal(instanceNameList[i])
+	}
+	baseSort(instanceNameList, getNameNOrdinalFunc, nil, true)
+	return instanceNameList, nil
+}
+
+func GenerateInstanceNamesFromTemplate(parentName, templateName string, replicas int32, offlineInstances []string, indexPoints []int32) []string {
+	instanceNames := GenerateInstanceNames(parentName, templateName, replicas, 0, offlineInstances, indexPoints)
 	return instanceNames
 }
 
 // GenerateInstanceNames generates instance names based on certain rules:
 // The naming convention for instances (pods) based on the Parent Name, InstanceTemplate Name, and ordinal.
 // The constructed instance name follows the pattern: $(parent.name)-$(template.name)-$(ordinal).
+// If the user specifies indexRanges, then only use indexRanges to generate instance names.
 func GenerateInstanceNames(parentName, templateName string,
-	replicas int32, ordinal int32, offlineInstances []string) ([]string, int32) {
-	usedNames := sets.New(offlineInstances...)
+	replicas int32, ordinal int32, offlineInstances []string, indexPoints []int32) []string {
 	var instanceNameList []string
+	if len(indexPoints) > 0 {
+		for _, indexPoint := range indexPoints {
+			var name string
+			if len(templateName) == 0 {
+				name = fmt.Sprintf("%s-%d", parentName, indexPoint)
+			} else {
+				name = fmt.Sprintf("%s-%s-%d", parentName, templateName, indexPoint)
+			}
+			instanceNameList = append(instanceNameList, name)
+		}
```

Comment:
The logic for handling offlineInstances is missing.

Reply:
As mentioned in the previous comment reply, the `IndexPoints` here have actually already been filtered through the `OfflineInstances`.